### DISCUSSION TITLE
Netbox Monitoring UPS by SNMP for Power utilization feature.  Script

### DISCUSSION THREAD

**Discussion by JHarrison712** [netbox-community/netbox#15860] [Q&A]
I made an SNMP script that sites on the netbox server and pushes every so often. Since the way our datacenter and other places are design we have Panel A > Power Feed A > UPS > PDU > Servers.

******************************************************************************************************************************************
```
from pysnmp.hlapi import *
from pynetbox import api

# NetBox API credentials and UPS device ID
NETBOX_URL = 'http://netbox.example.com/api/'
NETBOX_TOKEN = '[REDACTED]'
UPS_DEVICE_ID = 6 # Replace with the actual UPS device ID in NetBox

# SNMP parameters for the UPS device (SNMPv1)
SNMP_COMMUNITY = 'public'  # SNMP community string
SNMP_DEVICE = ('[REDACTED]', 161)  # Replace with the UPS IP address and SNMP port
# SNMP OID for UPS load
OID_UPS_LOAD = '1.3.6.1.2.1.33.1.4.4.1.5.1'

# Function to fetch UPS load using SNMP
def get_ups_load():
    errorIndication, errorStatus, errorIndex, varBinds = next(
        getCmd(SnmpEngine(),
               CommunityData(SNMP_COMMUNITY),
               UdpTransportTarget(SNMP_DEVICE, timeout=1, retries=5),  # Increase timeout and retries if needed
               ContextData(),
               ObjectType(ObjectIdentity(OID_UPS_LOAD)))
    )

    if errorIndication:
        print("SNMP Error:", errorIndication)
        return None
    elif errorStatus:
        print("SNMP Error:", errorStatus.prettyPrint())
        return None
    else:
        for varBind in varBinds:
            return int(varBind[1])

# Function to update UPS load in NetBox
def update_ups_load_in_netbox(upl):
    try:
        nb = api(url=NETBOX_URL, token=NETBOX_TOKEN)
        device = nb.dcim.devices.get(UPS_DEVICE_ID)
        device.custom_fields['ups_load'] = upl
        device.save()
        print("UPS load updated in NetBox:", upl)
    except Exception as e:
        print("Error updating UPS load in NetBox:", e)

# Main function
def main():
    ups_load = get_ups_load()
    if ups_load is not None:
        update_ups_load_in_netbox(ups_load)

if __name__ == "__main__":
    main()

```
***************************************************************************************************************************************
I worked out all the kinks and the only issues I have now is "Error: Device not found in NetBox."

Even tho when I go to my Netbox that UPS it shows DCIM.Devices.6

**Comment by markkuleinio** [comment]
Please separate your whole codeblock from other text with three backticks (```) to format it nicely. (Use the preview tab to verify before posting)

I'm confused because I don't think that code could ever print error "Error: Device not found in NetBox." Did I misread the code (due to uncomfortable formatting) or did you mean something else?

**Reply by JHarrison712** [reply]
Fixed sorry about that

**Reply by markkuleinio** [reply]
Thanks, can you show the actual error message you get?

**Reply by JHarrison712** [reply]
![2024-04-26 08_55_11-LWIS-VM-W10 - VMware Workstation](https://github.com/netbox-community/netbox/assets/6063493/af6f3c1e-7e27-40e5-a1aa-200385789f4b)

**Reply by JHarrison712** [reply]
![ups device](https://github.com/netbox-community/netbox/assets/6063493/6383874e-92c8-4e6b-ad6e-a6918c8c3555)

**Reply by JHarrison712** [reply]
I even changed it to go to power port and changed the ID and created the custom field

**Reply by markkuleinio** [reply]
What happens if you fix the NetBox URL by removing the /api/ suffix? `pynetbox.api()` takes the base URL.

**Reply by markkuleinio** [reply]
In any case, when having problems with Python code, masking the exception by catching `Exception` makes troubleshooting hard. Just remove catching the exception and let it print out the full traceback.

**Reply by markkuleinio** [reply]
I didn't find the string "Device not found in NetBox" in either NetBox or `pynetbox` code.

**Reply by JHarrison712** [reply]
hmm odd, maybe the script doesn't work then.

**Reply by markkuleinio** [reply]
What's your guess where that error message is coming from? Are you really connecting directly to NetBox?

**Reply by JHarrison712** [reply]
Let me go back to that snapshot and try again without an api key and see what errors come back.

**Reply by JHarrison712** [reply]
I get this now. 
![image](https://github.com/netbox-community/netbox/assets/6063493/8776eea1-df69-4ac2-a270-73d4e9a18b9b)

**Reply by markkuleinio** [reply]
Now that's a real error, meaning that `nb.dcim.devices.get()` returned `None`.

Does the user (whose token it is) have permission to view that device?

**Reply by JHarrison712** [reply]
After I modified these > 
![image](https://github.com/netbox-community/netbox/assets/6063493/d5946727-7ec7-4466-a1df-d765ca310903)
I now get 
![image](https://github.com/netbox-community/netbox/assets/6063493/4e021d48-d200-4f72-a93b-a5f7d6abe1ca)

**Reply by JHarrison712** [reply]
Yes I'm just using the admin account I have for right now.

**Reply by markkuleinio** [reply]
I still don't think that error is coming from either NetBox or `pynetbox`, so your environment is using something that you haven't revealed yet.

What are your NetBox and `pynetbox` versions?

How about removing try-except as I suggested?
